### PR TITLE
CORTX-31824: Activity framework Modifications

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Problem Statement
- [CORTX-31824](https://jts.seagate.com/browse/CORTX-31824)
- 1. Modify activity framework to accommodate new fields in the model-> 
name, created_time, updated_time etc.
- 2. Manage the activity 'status' internally and add a new field 'status_description' which will describe more about 'status'.
- 3. Activity ID should be generated using uuid.
- 4. Expose suspend interface

# Design
-  For Bug describe the fix here. 
-  For Feature, Post the link to the solution page on the confluence CORTX Foundation Library 

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing
- Are test cases updated along with code changes due to Enhancements/Bugs [Y/N]: Y
- Confirm that new test cases are added to regression and sanity plan files and relevant feature plan files [Y/N]:
- Confirm that Test Cases are added for new features added [Y/N]: Y
- Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- Confirm Testing was performed with installed RPM/K8s deployment [Y/N]:  Y

# Review Checklist 
####  Before posting the PR please ensure:
- [x] PR is self reviewed
- [x] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [ ] New package/s added to setup.py?
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [x] Verification needs to be done before marked as Closed/Verified.

# KVstore/Confstore feature (changes/fixes) checklist
#### Confirm changes are made for:
- [ ] ConfStore
- [ ] KVStore
- [ ] ConfCli
- [ ] Test cases added for all above 3
- [ ] changes done in all the KVpayloads (ConsulKvPayload, IniKvPayload, KvPayload)

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Review by abhijit1patil — APPROVED** [review]

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 1]
```diff
@@ -19,30 +19,51 @@
...
 from cortx.utils.kv_store import KvPayload
 from cortx.utils.activity_tracker.error import ActivityError
+from cortx.utils.activity_tracker import const
+
+class Status(Enum):
+
+    NEW = const.NEW
+    IN_PROGRESS = const.IN_PROGRESS
+    COMPLETED = const.COMPLETED
+    FAILED = const.FAILED
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:36]: Valid states are NEW, IN_PROGRESS, COMPLETED, SUSPENDED

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 2]
```diff
@@ -19,30 +19,51 @@
...
-                raise ActivityError(errno.EINVAL, "Missing resource_path/description")
-
-            self._id = f'{resource_path}>%s' %time.time()
-            self._payload['id'] = self._id
-            self._payload['resource_path'] = resource_path
-            self._payload['description'] = description
-            self._payload['pct_complete'] = 0
+            activity_name = kwargs.get(const.NAME)
+            resource_path = kwargs.get(const.RESOURCE_PATH)
+            description = kwargs.get(const.DESCRIPTION)
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:54]: What is any of the kwargs is missing ? Did you test ? Do we raise exception, handle it ?

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 3]
```diff
@@ -19,30 +19,51 @@
...
+            if activity_name is None or resource_path is None or description is None:
+                raise ActivityError(errno.EINVAL, "Missing activity_name/resource_path/description")
+
+            self._id = uuid.uuid1().hex
+            self._payload[const.ID] = self._id
+            self._payload[const.NAME] = activity_name
+            self._payload[const.RESOURCE_PATH] = resource_path
+            self._payload[const.DESCRIPTION] = description
+            self._payload[const.PROGRESS] = 0
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:63]: self._payload[const.PCT_COMPLETE] = 0

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 4]
```diff
@@ -59,17 +80,19 @@ def payload(self) -> dict:
     def set_attr(self, attr: str, val: str):
         self._payload[attr] = val
 
-    def set_status(self, pct_complete: int, status: str):
-        self._payload['pct_complete'] = pct_complete
-        self._payload['status'] = status
+    def set_status(self, progress: int, status: str, status_description: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:83]: I think this interface should be modified to 
```
def set_progress(self, pct_complete: int, description: str):
  self._payload[const.STATUS] = IN_PROGRESS
  ...
```

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 5]
```diff
@@ -59,17 +80,19 @@ def payload(self) -> dict:
...
+        self._payload[const.PROGRESS] = progress
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:84]: Also need to add an interface name 'suspend'. Refer to the state machine

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 6]
```diff
@@ -99,13 +128,24 @@ def start(activity: ActivityEntry):
...
+        Updates the activity progress, status and status_description.
+        Records the current time as the updated time.
+        """
+        if not isinstance(activity, ActivityEntry):
+            raise ActivityError(errno.EINVAL, "update(): Invalid arg %s", activity)
+        try:
+            status = Status(status).value
+        except ValueError:
+            raise ActivityError(errno.EINVAL, "update(): invalid status value %s", status)
+        activity.set_status(progress, status, status_description)
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:148]: Did you test if search is working ?

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 7]
```diff
@@ -99,13 +128,24 @@ def start(activity: ActivityEntry):
...
+        if not isinstance(activity, ActivityEntry):
+            raise ActivityError(errno.EINVAL, "finish(): Invalid arg %s", activity)
         activity.finish()
         Activity._kv_store.set([activity.id], [activity.payload.json])
 
     @staticmethod
-    def update(activity: ActivityEntry, pct_complete: int, status: str):
-        """Updates the activity progress."""
-        activity.set_status(pct_complete, status)
+    def update(activity: ActivityEntry, progress: int, status: str, status_description: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:137]: `def update(activity: ActivityEntry, pct_complete: int, description: str):`

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 8]
```diff
@@ -82,9 +105,15 @@ def init(backend_url):
         Activity._kv_store = KvStoreFactory.get_instance(backend_url)
 
     @staticmethod
-    def create(resource_path: str, description: str):
-        """Creates a activity for the given resource."""
-        activity = ActivityEntry(resource_path=resource_path, description=description)
+    def create(name:str, resource_path: str, description: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:108]: Did you test this with say Json or consul backend ?

File: py-utils/test/activity_tracker/test_activity_store.py [hunk 9]
```diff
@@ -51,18 +51,18 @@ class TestActivityStore(unittest.TestCase):
     def test_activity_store_create_activity(self):
         """Test activity creation."""
         Activity.init(sample_file_url)
-        activity = Activity.create("Activities", "Activity Status")
+        activity = Activity.create("Activity","Activities", "Activity Status")
```
Comment (ujjwalpl) [inline py-utils/test/activity_tracker/test_activity_store.py:54]: space after comma

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 8 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:108]: Tested with json and consul both. CSM CRUD APIs are using consul backend.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 6 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:148]: Search is not working. Im going to work on search interface in next sprint.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 8 above)
Comment (ujjwalpl) [reply py-utils/src/utils/activity_tracker/activity_store.py:108]: Search may not work with consul. Hence its important to confirm that it works.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 6 above)
Comment (ujjwalpl) [reply py-utils/src/utils/activity_tracker/activity_store.py:148]: Ok

File: py-utils/test/activity_tracker/test_activity_store.py [hunk 10]
```diff
@@ -51,18 +51,18 @@ class TestActivityStore(unittest.TestCase):
...
         try:
-            activity_id = activity.id.split('>')[1]
+            activity_id = activity.id
         except:
             rc = 1
             raise ActivityError(errno.EINVAL, "Error: While fetching Activity ID")
         self.assertEqual(rc, 0)
         with self.assertRaises(ActivityError):
-            Activity.create("activity", None)
+            Activity.create("Activity","Activities", None)
```
Comment (ujjwalpl) [inline py-utils/test/activity_tracker/test_activity_store.py:63]: Space after comma

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 2 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:54]: ActivityEntry will be called through Activity.create() interface, there we have ensured this.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 8 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:108]: sure

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 3 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:63]: As per discussion, I have changed this from 'pct_complete' -> progress. 
do you want me to change it to ''pct_complete' ?

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 4 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:83]: yes, currently, user can provide the 'status' value too, that is why Im accepting status. 
Im going to work on this, once we finalize the design for the update interface in next sprint for different scenarios.
eg. For updating the status to "SUSPENDED", user needs to provide status value too.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 5 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:84]: sure

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 7 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:137]: yes, currently, user can provide the 'status' value too, that is why Im accepting status.
Im going to work on this, once we finalize the design for the update interface in next sprint for different scenarios.
eg. For updating the status to "SUSPENDED", user needs to provide status value too.

File: py-utils/test/activity_tracker/test_activity_store.py (hunk 9 above)
Comment (rohitkolapkar) [reply py-utils/test/activity_tracker/test_activity_store.py]: updated

File: py-utils/test/activity_tracker/test_activity_store.py (hunk 10 above)
Comment (rohitkolapkar) [reply py-utils/test/activity_tracker/test_activity_store.py]: updated

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 2 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:53]: > What if any of the kwargs is missing ? Did you test ? Do we raise exception, handle it ?

Hi @ujjwalpl, this is handled, if the key is missing, the value would be None and I can see None check is already applied, we also raise an exception.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 3 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: > self._payload[const.PCT_COMPLETE] = 0

As we discussed, changed it to const.PCT_PROGRESS

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 2 above)
Comment (ujjwalpl) [reply py-utils/src/utils/activity_tracker/activity_store.py:53]: Every module is responsible for own validations unless its completely internal module and not exposed to anything

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 3 above)
Comment (ujjwalpl) [reply py-utils/src/utils/activity_tracker/activity_store.py]: **pct_complete** or **pct_progress**. 
We need to be more meaningful. progress can be 0 to 10000. pct_progress means 0 to 100.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Comment by rohitkolapkar** [comment]
@kupranay please merge this PR to feature branch. All the suggested changes will be taken care in https://jts.seagate.com/browse/CORTX-32261

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 7 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: updated

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 5 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: added

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 3 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: yes.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 4 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 11]
```diff
@@ -59,17 +70,24 @@ def payload(self) -> dict:
     def set_attr(self, attr: str, val: str):
         self._payload[attr] = val
 
-    def set_status(self, pct_complete: int, status: str):
-        self._payload['pct_complete'] = pct_complete
-        self._payload['status'] = status
+    def set_progress(self, pct_progress: int, status_desc: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:73]: We need to track start time for the activity.

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 12]
```diff
@@ -82,30 +100,61 @@ def init(backend_url):
...
+        """
+        activity = ActivityEntry(name=name, resource_path=resource_path,
+            description=description)
         Activity._kv_store.set([activity.id], [activity.payload.json])
         return activity
 
     @staticmethod
-    def start(activity: ActivityEntry):
-        """Start the activity. Records the current time as the start time."""
+    def update(activity: ActivityEntry, pct_progress: int, status_desc: str = const.IN_PROGRESS_DESC):
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:116]: Pls use less than 80 chars

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 13]
```diff
@@ -59,17 +70,24 @@ def payload(self) -> dict:
...
+        if pct_progress < 0 or pct_progress > 99:
+            raise ActivityError(errno.EINVAL, "update(): Invalid pct_progress: %s", pct_progress)
+        self._payload[const.PCT_PROGRESS] = pct_progress
+        self._payload[const.STATUS] = const.IN_PROGRESS
+        self._payload[const.STATUS_DESC] = status_desc
+        self._payload[const.UPDATED_TIME] = int(time.time())
 
-    def start(self):
-        self._payload['start_time'] = time.time()
+    def finish(self, status_desc: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/activity_tracker/activity_store.py:81]: you need to accept return code. Please refer to the state machine.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 14]
```diff
@@ -82,30 +100,61 @@ def init(backend_url):
...
+        Updates the pct_progress and status_description.
+        Sets the status to IN_PROGRESS.
+        Records current time as the updated_time.
+        """
         if not isinstance(activity, ActivityEntry):
-            raise ActivityError(errno.EINVAL, "start(): Invalid arg %s", activity)
-        activity.start()
+            raise ActivityError(errno.EINVAL, "update(): Invalid arg %s", activity)
+        activity_data = json.loads(activity.payload.json)
+        if activity_data.get(const.STATUS) == const.COMPLETED:
```
Comment (mssawant) [inline py-utils/src/utils/activity_tracker/activity_store.py:125]: [suggestion] this condition is redundant, instead you can have `activity.is_complete(activity_data)`

Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 12 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 11 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:76]: Hi, as we have discussed, create and start is one and the same. so when an activity is created, we are recording current_time as created_time.

And when we call update, suspend or finish, we are recording current_time as updated_time.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 13 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

**Review by ujjwalpl — APPROVED** [review]

**Review by suryakumar1024 — APPROVED** [review]

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 15]
```diff
@@ -59,17 +73,26 @@ def payload(self) -> dict:
     def set_attr(self, attr: str, val: str):
         self._payload[attr] = val
 
-    def set_status(self, pct_complete: int, status: str):
-        self._payload['pct_complete'] = pct_complete
-        self._payload['status'] = status
+    def set_progress(self, pct_progress: int, status_desc: str):
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/activity_tracker/activity_store.py:76]: changes in method signature.

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 16]
```diff
@@ -59,17 +73,26 @@ def payload(self) -> dict:
...
+            raise ActivityError(errno.EINVAL, f"update(): Invalid pct_progress:\
+                {pct_progress}")
+        self._payload[const.PCT_PROGRESS] = pct_progress
+        self._payload[const.STATUS] = const.IN_PROGRESS
+        self._payload[const.STATUS_DESC] = status_desc
+        self._payload[const.UPDATED_TIME] = int(time.time())
 
-    def start(self):
-        self._payload['start_time'] = time.time()
+    def finish(self, rc: int, status_desc: str):
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/activity_tracker/activity_store.py:85]: changes in method signature.

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 17]
```diff
@@ -82,30 +105,69 @@ def init(backend_url):
         Activity._kv_store = KvStoreFactory.get_instance(backend_url)
 
     @staticmethod
-    def create(resource_path: str, description: str):
-        """Creates a activity for the given resource."""
-        activity = ActivityEntry(resource_path=resource_path, description=description)
+    def create(name:str, resource_path: str, description: str) -> ActivityEntry:
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/activity_tracker/activity_store.py:108]: changes in method signature.

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 18]
```diff
@@ -82,30 +105,69 @@ def init(backend_url):
...
+            raise ActivityError(errno.EINVAL, "update(): pct_progress: %s can \
+                not be less than the previously updated value", pct_progress)
+        activity.set_progress(pct_progress, status_desc)
         Activity._kv_store.set([activity.id], [activity.payload.json])
 
     @staticmethod
-    def finish(activity: ActivityEntry):
-        """Completes a activity. Records current time as the completion time."""
-        activity.finish()
+    def finish(activity: ActivityEntry, rc: int, status_desc: str = \
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/activity_tracker/activity_store.py:142]: changes in method signature.

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 19]
```diff
@@ -82,30 +105,69 @@ def init(backend_url):
...
+            raise ActivityError(errno.EINVAL, "finish(): Invalid arg %s",
+                activity)
+        activity.finish(rc, status_desc)
         Activity._kv_store.set([activity.id], [activity.payload.json])
 
     @staticmethod
-    def update(activity: ActivityEntry, pct_complete: int, status: str):
-        """Updates the activity progress."""
-        activity.set_status(pct_complete, status)
+    def suspend(activity: ActivityEntry, status_desc: str = \
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/activity_tracker/activity_store.py:156]: changes in method signature.

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 20]
```diff
@@ -133,10 +195,15 @@ def search(resource_path: str, filters: list) -> list:
         return out_list
 
     @staticmethod
-    def get(activity_id: str):
-        """Gets the activity details."""
+    def get(activity_id: str) -> ActivityEntry:
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/activity_tracker/activity_store.py:198]: changes in method signature.

**Review by rahul-tripathi-git — CHANGES_REQUESTED** [review]
There are changes in method signatures. Please answer following questions:
1. Which component/s will be affected by these changes.
2. If there are associated PR/s in other repos please mention them.

Also please make sure that the test cases added are added to sanity and other relevant plan files. (it will include these test cases in utils test framework.)

**Comment by rohitkolapkar** [comment]
> There are changes in method signatures. Please answer following questions:
> 
> 1. Which component/s will be affected by these changes.
> 2. If there are associated PR/s in other repos please mention them.
> 
> Also please make sure that the test cases added are added to sanity and other relevant plan files. (it will include these test cases in utils test framework.)

1 and 2 -> Component/s will not be affected as No one is using it. CSM is going to use it in upcoming sprint.

Added test pln file, aslo added in sanity and regression pln.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 15 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:76]: No component will be affected as No one is using it. Method signature change is a part of enhancement to framework.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 16 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:85]: No component will be affected as No one is using it. Method signature change is a part of enhancement to framework.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 17 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:108]: No component will be affected as No one is using it. Method signature change is a part of enhancement to framework.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 18 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:142]: No component will be affected as No one is using it. Method signature change is a part of enhancement to framework.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 19 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:156]: This is not a method signature change, suspend is a new interface.

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 20 above)
Comment (rohitkolapkar) [reply py-utils/src/utils/activity_tracker/activity_store.py:198]: No method signature change,  this method has been returning  ActivityEntry, added code readability.

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Pull request merged**